src/FieldMapB.cpp

```diff
@@ -0,0 +1,377 @@
+  void Configure(std::vector<xml_comp_t> dimensions);
+  void LoadMap(const std::string& map_file, double scale);
+  bool GetIndices(double R, double Z, int *idxR, int *idxZ, double *deltaR, double *deltaZ);
+  bool GetIndices(double X, double Y, double Z, int *idxX, int *idxY, int *idxZ, double *deltaX, double *deltaY, double *deltaZ);
+  void SetTransform(const Transform3D& tr)
+  {
+    trans     = tr;
+    trans_inv = tr.Inverse();
+  }
+
+  virtual void fieldComponents(const double* pos, double* field);
+
+private:
+  bool                                                          case_BrBz; // is it BrBz or not (BxByBz)
+  Transform3D                                                   trans, trans_inv; // transformation matrices
+  std::vector<double>                                           steps, mins, maxs; // B map cell info
+  int                                                           ir, ix, iy, iz; // lookup indices
+  double                                                        ir_d, ix_d, iy_d, iz_d; // transient indices for std::modf
+  double                                                        dr, dx, dy, dz; // deltas for interpolation
+  std::vector<std::vector<std::vector<double>>>                 Bvals_RZ; // B map values:  {R}, {Z}, {Br,Bz}
+  std::vector<std::vector<std::vector<std::vector<double>>>>    Bvals_XYZ; // B map values: {X}, {Y}, {Z}, {Bx,By,Bz}
+  ROOT::Math::XYZPoint                                          B_interpolated; // interpolated B
+};
+
+// constructor
+FieldMapB::FieldMapB(const std::string& field_type_str, const std::string& coord_type_str)
+{
+  std::string ftype = field_type_str;
+  for (auto& c : ftype) {
+    c = tolower(c);
+  }
+
+  // set type
+  if (ftype == "magnetic") {
+    DD4HEP_FIELD_TYPE = CartesianField::MAGNETIC;
+  } else if (ftype == "electric") {
+    DD4HEP_FIELD_TYPE = CartesianField::ELECTRIC;
+  } else {
+    DD4HEP_FIELD_TYPE = CartesianField::UNKNOWN;
+    std::cout << "FieldMapB Warning: Unknown field type " << field_type_str << "!" << std::endl;
+  }
+
+  if( coord_type_str.compare("BrBz") == 0 ) { // BrBz
+    case_BrBz = true;
+  }
+  else { // BxByBz
+    case_BrBz = false;
+  }
+}
+
+// fill field vector
+void FieldMapB::Configure(std::vector<xml_comp_t> dimensions)
+{
+  // Fill vectors with step size, min, max for each dimension
+  for( auto el : dimensions ) {
+    steps.push_back( el.step() );
+    mins.push_back( getAttrOrDefault<double>(el, _Unicode(min), 0) );
+    maxs.push_back( getAttrOrDefault<double>(el, _Unicode(max), 0) );
+  }
+
+  if( case_BrBz ) {
+    int nr = int((maxs[0] - mins[0]) / steps[0]) + 2;
+    int nz = int((maxs[1] - mins[1]) / steps[1]) + 2;
+
+    Bvals_RZ.resize(nr);
+    for (auto& B2 : Bvals_RZ) {
+      B2.resize(nz);
+      for (auto& B : B2) {
+        B.resize(2, 0.); // Br, Bz
+      }
+    }
+  }
+  else {
+    int nx = int((maxs[0] - mins[0]) / steps[0]) + 2;
```

Comment:
Is this not wrong or at least a bit fragile since it depends on truncation of potentially incorrect floating point representation? If `maxs = mins + steps` then we naively get `int(1) + 2 = 3` when we should get a size of 2 for the 2 grid points. Overcount is better than undercount when prepping for a resize, but can we fix this with roundf, remquof, or use a modf with iptr arg?

Reply:
I now use remquof and with it, I removed the overcounting: +2 --> +1.